### PR TITLE
Implemented extra parachain host runtime API calls

### PR DESCRIPTION
**Pull request by kishansagathiya**
Implemented following parachain host runtime calls
- ParachainHost_persisted_validation_data
- ParachainHost_validation_code

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Primary Reviewer

@kanishkatn

### CODE REVIEW

File: lib/runtime/wasmer/exports_test.go [hunk 1]
```diff
@@ -1214,6 +1214,47 @@ func TestInstance_GrandpaSubmitReportEquivocationUnsignedExtrinsic(t *testing.T)
...
+	rt := NewTestInstanceWithTrie(t, runtime.WESTEND_RUNTIME_v0942, tt)
+
+	parachainID := uint32(1000)
+	assumption := parachaintypes.NewOccupiedCoreAssumption()
+	err := assumption.Set(parachaintypes.IncludedOccupiedCoreAssumption{})
+	require.NoError(t, err)
+
+	validationCode, err := rt.ParachainHostValidationCode(parachainID, assumption)
+	require.NoError(t, err)
+	require.NotEmpty(t, validationCode)
```
Comment (kanishkatn) [inline lib/runtime/wasmer/exports_test.go:1255]: Let's also add a check to ensure it matches the validation code that we're setting. There was an issue with it and I am not sure if it is fixed.

File: lib/runtime/wasmer/exports.go [hunk 2]
```diff
@@ -347,6 +347,85 @@ func (in *Instance) GrandpaSubmitReportEquivocationUnsignedExtrinsic(
...
+		return nil, err
+	}
+
+	var validationCode []byte
+	err = scale.Unmarshal(encodedValidationCode, &validationCode)
+	if err != nil {
+		return nil, fmt.Errorf("scale decoding: %w", err)
+	}
+
+	return (*parachaintypes.ValidationCode)(&validationCode), nil
```
Comment (kanishkatn) [inline lib/runtime/wasmer/exports.go:426]: I think we can directly unmarshal into `ValidationCode` and avoid using `[]byte`. I was just playing around with it and had committed this code.

File: lib/parachain/types/types.go [hunk 3]
```diff
@@ -389,3 +389,83 @@ func NewCandidateEvents() (scale.VaryingDataTypeSlice, error) {
...
+
+// NewOccupiedCoreAssumption creates a OccupiedCoreAssumption varying data type.
+func NewOccupiedCoreAssumption() OccupiedCoreAssumption {
+	vdt, err := scale.NewVaryingDataType(
+		IncludedOccupiedCoreAssumption{},
+		FreeOccupiedCoreAssumption{},
+		TimedOutOccupiedCoreAssumption{})
+	if err != nil {
+		panic(err)
+	}
```
Comment (axaysagathiya) [inline lib/parachain/types/types.go:468]: You can use `scale.MustNewVaryingDataType()` here.
```suggestion
	vdt := scale.MustNewVaryingDataType(
		IncludedOccupiedCoreAssumption{},
		FreeOccupiedCoreAssumption{},
		TimedOutOccupiedCoreAssumption{})
```

File: lib/parachain/types/types_test.go [hunk 4]
```diff
@@ -371,3 +371,46 @@ func TestMustHexTo32BArray(t *testing.T) {
...
+	require.NoError(t, err)
+	require.Equal(t, expected, actual)
+
+	newpvd := PersistedValidationData{}
+	err = scale.Unmarshal(actual, &newpvd)
+	require.NoError(t, err)
+	require.Equal(t, pvd, newpvd)
+}
+
+func TestOccupiedCoreAssumption(t *testing.T) {
```
Comment (axaysagathiya) [inline lib/parachain/types/types_test.go:395]: also test for unmarshal.
you can refer [TestCollationProtocol](https://github.com/ChainSafe/gossamer/blob/cae09d69e2c388851ed92d1aa3eedcef1dadebe9/lib/parachain/collation_protocol_test.go#L29-L147)

**Review by axaysagathiya — APPROVED** [review]

**Pull request merged**